File: Python/Day 7/AOC22_7B.py
import sys, copy, re
from collections import defaultdict as dd
read = sys.stdin.read

f = open("AOC22_7A_in.txt")
inp = f.read().split('\n')

ans = 0

path = []
sz = dd(int)

"""

if cd, then we go into a folder or come out
if we go in, then everything inside

/ -> (a, b, c)

add b and c to the path

If we go into folder, add that folder to the path, ls
If ls, ignore anything that's a dir, add anything that isn't to everything on the current path
If we come out of folder, remove top node from path

"""

for line in inp:
    x = line.split()
    if x[0] == "$" and x[1] == "cd":
        if x[2] == "..":
            path.pop()
        else:
            path.append(x[2])
        continue
    if x[0] == "$":
        assert x[1] == "ls"
        continue
    if x[0] == "dir":
        continue
    for i in range(len(path)):
        sz[''.join(path[:i + 1])] += int(x[0])

allowed = 70000000
req = 30000000
used = sz["/"]
unused = allowed - used
rem = req - unused

ans = 10**10

for key, val in sz.items():
    if val >= rem:
        ans = min(ans, val)

print(ans)

File: C++/Day 7/AOC22_7B.cpp
#include <bits/stdc++.h>
#define print(x) std::cout << x << std::endl
using namespace std;

int main()
{
    ifstream InputFile("AOC22_7A_in.txt");
	
	vector<string> input;
	
    for(string line; getline(InputFile,line);)
    {
        input.push_back(line);
    }
	
	vector<string> path;
	map<string, int> sz;
    
	for(auto line: input)
    {
		while (line.back() == '\r') {
			line.pop_back();
		}
        if (line.substr(2,2)=="cd")
        {
            if (line.substr(5, line.size())=="..")
            {
                path.pop_back();
            }

            else if (path.empty())
			{
				path.push_back(".");
			}
			else
            {
                string name = line.substr(5, line.size());
				path.push_back(path.back() + "/" + name);
            }
        }
		else if (line.substr(2,2) == "ls")
		{
			continue;
		}
		else if (line[0] == 'd')
		{
			continue;
		}
		else
		{
			string num = "";
			for(char c: line) {
				if (c == ' ') break;
				num += c;
			}
			for(string S: path)
			{
				sz[S] += stoi(num);
			}
		}
    }
	int ans = 1e9;
	int rem = sz["."] - 40000000;
	for(auto [key, val] : sz)
	{
		if (val >= rem) {
			ans = min(ans, val);
		}
	}
	print(ans);
}